### DISCUSSION TITLE
I have problem with Adam optimizer.

### DISCUSSION THREAD

**Discussion by seoggu** [mrdbourke/pytorch-deep-learning#328] [Q&A] [answered]
Well, I'm recently following [PyTorch for Deep Learning in 2023: Zero to Mastery].

And for CNN model, I tried Adam optimizer instead of SGD.

when I use SGD, my model works great. But with Adam, my model seems like never learns anything. 

I heard that Adam optimizer is the most powerful optimizer. So I really want to use it. 

should I change architecture of my model to use Adam?

**Comment by pusapatiakhilraju** [comment]
Hey Seoggu, 

It is not necessary that ADAM will work well in all the conditions. From my understanding ADAM is used to converge but doesn't generalise well. 
here is a link to the article that can help you understand this further. https://towardsai.net/p/l/why-should-adam-optimizer-not-be-the-default-learning-algorithm
Try playing around with the learning rate and architecture a little to see if that helps.

Also, I have used ADAM to do the Pytorch classification section in the Pytorch course. 
Here is the link to my github profile
https://github.com/pusapatiakhilraju/Pytorch-Neural-Net-Classification/blob/master/Neural-Net-Classifcation.ipynb
 you can see in cell 93 that I have used ADAM and the results in cell number 97 are good.
 
 Hope this answers your question.

**Reply by seoggu** [reply]
wll i tried some manipulation to my model.

First, i simply changed my optimizer sgd to adam. 

here is architecture of my model
conv2d
relu
conv2d
relu
maxpool2d

conv2d
relu
conv2d
relu
maxpool2d

flatten
linear

This model never learns with adam

then i removed all ReLU layer, adam works well.

I tried model with layer stack of linear + relu and this one also works fine.

Is there any relations between adam and convolutional layer or pooling layer?

**Comment by pusapatiakhilraju** [comment, marked as answer]
From my understanding there is no such relation relation between convolution layer, pooling layers and ADAM. I feel it's got to do with hyper parameters of your model. Play around with the hyper parameters of the ADAM optimiser. Change the learning rate, Beta(exponential decay rate) .

**Comment by mrdbourke** [comment]
Hi @seoggu,

@pusapatiakhilraju is right, there is no 100% set optimizer to use for each neural network.

Adam/SGD will be most of what you need for many neural networks, so I would recommend experimenting with each of these to begin with (try using different learning rates, as this is the most important hyperparameter when training neural networks).

However, once you start training larger models with more sophisticated training setups, you may want to try other optimizers: https://pytorch.org/docs/stable/optim.html

**Comment by seoggu** [comment]
@mrdbourke 
@pusapatiakhilraju 

I really appreciate both of you guys. 

I changed learning rate 0.1 to 0.0001. and finally my Adam optimizer works. Even better than SGD.

Adam is more accurate than SGD, but it seems like way slower than SGD. 

Anyway problem solved :)

**Reply by darkar18** [reply]
why did you change lr to .0001(any specific reasons?), did you try .001, does it speed up?

**Reply by PastorJordi** [reply]
Indeed it will be slower with a 1e3 fold smaller lr. @darkar18 using a "too large" lr may impair model's learning because weights adjustment in each backprop will be too big. Then, the model will never converge as a consequence.
You can try [decreasing lr using a scheduler](https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate) so it "goes fast" during the first epochs and "slow" in the final ones, once it almost has converged.
Disclaimer: I am a complete pytorch noob, but thit was my approach back when I used Keras 
I think [this one from Ruder is a good read](https://www.ruder.io/optimizing-gradient-descent/) to understand optimizer's nuances (It actually took me a while to find it due to the insane amount of plagiarized medium crap out there)
